src/stream.rs

```diff
@@ -0,0 +1,194 @@
+// Stream consumer and producer for ABCI Messages
+use integer_encoding::VarInt;
+use messages::abci::*;
+use mockstream::SharedMockStream;
+use protobuf::Message;
+use std::fmt::{self, Debug, Formatter};
+use std::io;
+use std::io::{Read, Write};
+use std::net::TcpStream;
+
+/// Wraps a stream for easier testing.  Original code from tomtau
+pub enum StreamWrapper {
+    Mocked(SharedMockStream),
+    Tcp(TcpStream),
+}
+
+impl io::Read for StreamWrapper {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        match *self {
+            StreamWrapper::Mocked(ref mut s) => s.read(buf),
+            StreamWrapper::Tcp(ref mut s) => s.read(buf),
+        }
+    }
+}
+
+impl io::Write for StreamWrapper {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        match *self {
+            StreamWrapper::Mocked(ref mut s) => s.write(buf),
+            StreamWrapper::Tcp(ref mut s) => s.write(buf),
+        }
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
+        match *self {
+            StreamWrapper::Mocked(ref mut s) => s.flush(),
+            StreamWrapper::Tcp(ref mut s) => s.flush(),
+        }
+    }
+}
+
+/// An ABCI Stream
+pub struct AbciStream {
+    stream: StreamWrapper,
+}
+
+impl Debug for AbciStream {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self.stream {
+            StreamWrapper::Mocked(_) => Ok(f.debug_struct("SharedMockStream").finish()?),
+            StreamWrapper::Tcp(ref s) => s.fmt(f),
+        }
+    }
+}
+
+impl AbciStream {
+    /// Write an ABCI specifc response to the stream.   Properly sizes the output
+    /// based on the size of the response.
+    pub fn write_response(&mut self, response: &Response) -> io::Result<()> {
+        let msg_size = response.compute_size() as usize;
+        let varint = i64::encode_var_vec(msg_size as i64);
+        let mut output = Vec::<u8>::with_capacity((msg_size + varint.len()) as usize);
```

Comment:
can't we write to stream directly?

Reply:
Not sure what you're asking.  Do you mean why the `write_response` method vs just doing it in the `respond()` ?  If so, it's because I wanted the logic for `mock` and `tcpstream` in the same place

Reply:
No, I meant doing smth like `self.stream.write_all(varint.as_slice()); self.stream.write_all(response.as_slice())`

Reply:
@melekes `response` doesn't appear to have an `as_slice()`